File: PSME/application/src/rest/endpoints/fabric/port_metrics.cpp
/*!
 * @brief Port metrics endpoint
 *
 * @copyright Copyright (c) 2017-2019 Intel Corporation.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at http://www.apache.org/licenses/LICENSE-2.0.
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 *
 * @file port_metrics.cpp
 */

#include "psme/rest/endpoints/fabric/port_metrics.hpp"



using namespace psme::rest;
using namespace psme::rest::constants;

namespace {
json::Json make_prototype() {
    json::Json r(json::Json::value_t::object);

    r[Common::ODATA_CONTEXT] = "/redfish/v1/$metadata#PortMetrics.PortMetrics";
    r[Common::ODATA_ID] = json::Json::value_t::null;
    r[Common::ODATA_TYPE] = "#PortMetrics.v1_0_0.PortMetrics";
    r[Common::ID] = json::Json::value_t::null;
    r[Common::NAME] = "Port Metrics";
    r[Common::DESCRIPTION] = "Fabric Port Metrics";

    r[constants::Common::HEALTH] = json::Json::value_t::null;

    return r;
}
}


endpoint::PortMetrics::PortMetrics(const std::string& path) : EndpointBase(path) {}


endpoint::PortMetrics::~PortMetrics() {}


void endpoint::PortMetrics::get(const server::Request& request, server::Response& response) {
    auto port_uuid = psme::rest::model::find<agent_framework::model::Fabric, agent_framework::model::Switch, agent_framework::model::Port>(
        request.params).get_uuid();

    auto json = ::make_prototype();
    json[Common::ODATA_ID] = PathBuilder(request).build();
    json[Common::ID] = constants::Common::METRICS;
    json[Common::NAME] = "Port " + request.params[PathParam::PORT_ID] + " Metrics";

    utils::populate_metrics(json, port_uuid);

    set_response(response, json);
}

File: PODM/SW/pod-manager/podm-business-entities/src/main/java/com/intel/podm/business/entities/redfish/PortMetrics.java
/*
 * Copyright (c) 2017-2018 Intel Corporation
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package com.intel.podm.business.entities.redfish;

import com.intel.podm.business.entities.Eventable;
import com.intel.podm.business.entities.redfish.base.DiscoverableEntity;
import com.intel.podm.business.entities.redfish.base.Entity;
import com.intel.podm.common.types.Id;

import javax.persistence.Column;
import javax.persistence.Index;
import javax.persistence.JoinColumn;
import javax.persistence.OneToOne;
import javax.persistence.Table;
import java.util.Objects;

import static javax.persistence.CascadeType.MERGE;
import static javax.persistence.CascadeType.PERSIST;
import static javax.persistence.FetchType.LAZY;

@javax.persistence.Entity
@Table(name = "port_metrics", indexes = @Index(name = "idx_port_metrics_entity_id", columnList = "entity_id", unique = true))
@Eventable
public class PortMetrics extends DiscoverableEntity {
    @Column(name = "entity_id", columnDefinition = ENTITY_ID_STRING_COLUMN_DEFINITION)
    private Id entityId;

    @Column(name = "health")
    private String health;

    @OneToOne(fetch = LAZY, cascade = {MERGE, PERSIST})
    @JoinColumn(name = "port_id")
    private Port port;

    @Override
    public Id getId() {
        return entityId;
    }

    @Override
    public void setId(Id id) {
        this.entityId = id;
    }

    public String getHealth() {
        return health;
    }

    public void setHealth(String health) {
        this.health = health;
    }

    public Port getPort() {
        return port;
    }

    public void setPort(Port port) {
        if (!Objects.equals(this.port, port)) {
            unlinkPort(this.port);
            this.port = port;
            if (port != null && !this.equals(port.getPortMetrics())) {
                port.setPortMetrics(this);
            }
        }
    }

    public void unlinkPort(Port port) {
        if (Objects.equals(this.port, port)) {
            this.port = null;
            if (port != null) {
                port.unlinkMetrics(this);
            }
        }
    }

    @Override
    public void preRemove() {
        unlinkPort(port);
    }

    @Override
    public boolean containedBy(Entity possibleParent) {
        return isContainedBy(possibleParent, port);
    }
}
